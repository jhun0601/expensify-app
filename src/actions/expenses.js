import { v4 as uuidv4 } from "uuid";
import db from "../firebase/firebase";
import { push, ref, remove, onValue, get, child } from "firebase/database";

export const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense,
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "",
            note = "",
            amount = 0,
            createdAt = 0,
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        const postPush = push(ref(db));
        return push(ref(db, "expenses"), expense)
            .then((ref) => {
                dispatch(
                    addExpense({
                        id: ref.key,
                        ...expense,
                    })
                );
            })
            .catch((e) => {
                throw new Error(e);
            });
    };
};

export const removeExpense = ({ id } = {}) => ({
    type: "REMOVE_EXPENSE",
    id,
});

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => {
        return remove(ref(db, `expenses/${id}`)).then(() => {
            dispatch(removeExpense({ id }));
        });
    };
};
export const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates,
});

// SET EXPENSE
export const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses,
});
const dbRef = ref(db);

export const startSetExpenses = () => {
    return (dispatch) => {
        return get(child(dbRef, "expenses"))
            .then(
                (snapshot) => {
                    const expenses = [];
                    snapshot.forEach((childSnapshot) => {
                        expenses.push({
                            id: childSnapshot.key,
                            ...childSnapshot.val(),
                        });
                    });
                    dispatch(setExpenses(expenses));
                },
                { onlyOnce: true }
            )
            .catch((error) => {
                console.error(error);
            });
    };
};
