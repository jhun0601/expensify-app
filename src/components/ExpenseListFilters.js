import React from "react";

import "react-dates/initialize";
// import "react-dates/lib/css/_datepicker.css";
import { DateRangePicker } from "react-dates";

import { connect } from "react-redux";
import {
    setTextFilter,
    sortByDate,
    sortByAmount,
    setStartDate,
    setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: null,
        };

        this.onDatesChange = this.onDatesChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
    }

    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onSortChange = (e) => {
        if (e.target.value === "date") {
            this.props.sortByDate();
        } else if (e.target.value === "amount") {
            this.props.sortByAmount();
        }
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            className="text-input"
                            placeholder="Search expenses"
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            value={this.props.filters.sortBy}
                            className="select"
                            onChange={this.onSortChange}
                        >
                            <option value="amount">Amount</option>
                            <option value="date">Date</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                            startDateId={"3dfjklds"}
                            endDateId={"asdasf"}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters,
});
const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
});
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
