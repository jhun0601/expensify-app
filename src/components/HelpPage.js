import React from "react";

const HelpPage = () => <div>This is a Help Page.</div>;

export default HelpPage;
