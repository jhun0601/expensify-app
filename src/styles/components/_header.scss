.header {
    background: $dark-blue;
    &__content {
        display: flex;
        justify-content: space-between;
        align-items: center;
        padding: $s-size 0;
    }
    &__title {
        color: #fff;
        text-decoration: none;
        h1 {
            margin: 0;
        }
    }
}
