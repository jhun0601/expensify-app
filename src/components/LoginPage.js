import React from "react";
import { connect } from "react-redux";
import { startLogin } from "../actions/auth";
export const LoginPage = ({ startLogin }) => (
    <div className="box-layout">
        <div className="box-layout__container">
            <img
                src="/images/favicon.png"
                alt="expensify app"
                className="box-layout__logo"
            />
            <h1 className="box-layout__title">Expensify</h1>
            <p>It's time to get your expenses under control.</p>
            <button onClick={startLogin} className="button">
                Log in with Google
            </button>
        </div>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogin: () => dispatch(startLogin()),
});
export default connect(undefined, mapDispatchToProps)(LoginPage);
