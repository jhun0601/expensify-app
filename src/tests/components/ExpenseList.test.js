import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("Should render ExpenseList with expense", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should rendere expensesList with empty message", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});
