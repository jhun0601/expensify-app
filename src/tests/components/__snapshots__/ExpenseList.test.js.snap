// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with expense 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseItemList
    amount={100}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseItemList
    amount={1950}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseItemList
    amount={20000}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should rendere expensesList with empty message 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    No Expenses
  </p>
</div>
`;
